cmd/tsbs_generate_queries/databases/cratedb/devops.go

```diff
@@ -0,0 +1,214 @@
+// a set of column idents.
+//
+// For instance:
+//      max(cpu_time) AS max_cpu_time
+func (d *Devops) getSelectAggClauses(aggFunc string, idents []string) []string {
+	selectAggClauses := make([]string, len(idents))
+	for i, ident := range idents {
+		selectAggClauses[i] =
+			fmt.Sprintf("%[1]s(%[2]s) AS %[1]s_%[2]s", aggFunc, ident)
+	}
+	return selectAggClauses
+}
+
+// MaxAllCPU selects the MAX of all metrics under 'cpu' per hour for N random
+// hosts
+//
+// Queries:
+// cpu-max-all-1
+// cpu-max-all-8
+func (d *Devops) MaxAllCPU(qi query.Query, nHosts int) {
+	interval := d.Interval.RandWindow(devops.MaxAllDuration)
+	selectClauses := d.getSelectAggClauses("max", devops.GetAllCPUMetrics())
+
+	sql := fmt.Sprintf(`
+		SELECT
+			date_trunc('hour', ts) AS hour,
+			%s
+		FROM cpu
+		WHERE %s IN ('%s')
+		  AND ts >= %d
+		  AND ts < %d
+		GROUP BY hour
+		ORDER BY hour`,
+		strings.Join(selectClauses, ", "),
+		hostnameField,
+		strings.Join(d.GetRandomHosts(nHosts), "', '"),
+		interval.StartUnixMillis(),
+		interval.EndUnixMillis())
+
+	humanLabel := devops.GetMaxAllLabel("CrateDB", nHosts)
+	humanDesc := fmt.Sprintf("%s: %s", humanLabel, interval.StartString())
+	d.fillInQuery(qi, humanLabel, humanDesc, sql)
+}
+
+// GroupByTimeAndPrimaryTag selects the AVG of metrics in the group `cpu` per device
+// per hour for a day
+//
+// Queries:
+// double-groupby-1
+// double-groupby-5
+// double-groupby-all
+func (d *Devops) GroupByTimeAndPrimaryTag(qi query.Query, numMetrics int) {
+	metrics := devops.GetCPUMetricsSlice(numMetrics)
+	interval := d.Interval.RandWindow(devops.DoubleGroupByDuration)
+	selectClauses := d.getSelectAggClauses("mean", metrics)
+
+	sql := fmt.Sprintf(`
+		SELECT
+			date_trunc('hour', ts) AS hour,
+			%s
+		FROM cpu
+		WHERE ts >= %d
+		  AND ts < %d
+		GROUP BY hour, %s
+		ORDER BY hour`,
+		strings.Join(selectClauses, ", "),
+		interval.StartUnixMillis(),
+		interval.EndUnixMillis(),
+		hostnameField)
+
+	humanLabel := devops.GetDoubleGroupByLabel("CrateDB", numMetrics)
+	humanDesc := fmt.Sprintf("%s: %s", humanLabel, interval.StartString())
+	d.fillInQuery(qi, humanLabel, humanDesc, sql)
+}
+
+// GroupByOrderByLimit populates a query.Query that has a time WHERE clause,
+// that groups by a truncated date, orders by that date, and takes a limit:
+//
+// Queries:
+// groupby-orderby-limit
+func (d *Devops) GroupByOrderByLimit(qi query.Query) {
+	interval := d.Interval.RandWindow(time.Hour)
+	sql := fmt.Sprintf(`
+		SELECT
+			date_trunc('minute', ts) as minute,
+			max(usage_user)
+		FROM cpu
+		WHERE ts < %d
+		GROUP BY minute
+		ORDER BY minute DESC
+		LIMIT 5`,
+		interval.EndUnixMillis())
+
+	humanLabel := "CrateDB max cpu over last 5 min-intervals (random end)"
+	humanDesc := fmt.Sprintf("%s: %s", humanLabel, interval.EndString())
+	d.fillInQuery(qi, humanLabel, humanDesc, sql)
+}
+
+// LastPointPerHost finds the last row for every host in the dataset
+func (d *Devops) LastPointPerHost(qi query.Query) {
```

Comment:
not sure about this query

Reply:
This doesn't seem right to me. `DISTINCT` is global over all columns. 

We'd have to have some kind of GROUP BY?

Reply:
https://github.com/crate/tsbs/blob/19932e723e844ad33cff0ed73c4a994568dc75bc/cmd/tsbs_generate_queries/databases/influx/devops.go#L128

Reply:
it seems that it has to be smth like that 

```
SELECT * 
FROM (select tags['hostname'], max(ts) as ts from cpu group by tags['hostname']) t, cpu 
WHERE t.ts = cpu. ts AND t.tags['hostname'] = cpu.tags['hostname'];
```